Remotion/ObjectBinding/Web/UI/Controls/Validation/BocListValidationFailureRepository.cs

```diff
@@ -0,0 +1,306 @@
+      {
+        var cellFailures = GetOrCreateListOfCellFailures(cellDefinition);
+        cellFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public void GetAllFailures (ICollection<BusinessObjectValidationFailure> failures, bool markAsHandled)
+      {
+        GetRowFailures(failures, markAsHandled);
+        GetAllCellFailuresForRow(failures, markAsHandled);
+      }
+
+      public void GetRowFailures (ICollection<BusinessObjectValidationFailure> failures, bool markAsHandled)
+      {
+        if (_rowFailures == null)
+          return;
+
+        foreach (var handleableRowFailure in _rowFailures.Where(e => !e.Handled))
+        {
+          failures.Add(handleableRowFailure.Failure);
+          if (markAsHandled)
+            handleableRowFailure.MarkAsHandled();
+        }
+      }
+
+      private void GetAllCellFailuresForRow (ICollection<BusinessObjectValidationFailure> failures, bool markAsHandled)
+      {
+        if (_cellFailures == null)
+          return;
+
+        foreach (var handleableRowFailure in _cellFailures.SelectMany(c => c.Value.Where(e => !e.Handled)))
+        {
+          failures.Add(handleableRowFailure.Failure);
+          if (markAsHandled)
+            handleableRowFailure.MarkAsHandled();
+        }
+      }
+
+      public void GetCellFailures (ICollection<BusinessObjectValidationFailure> failures, IBocColumnDefinitionWithValidationSupport cellDefinition, bool markAsHandled)
+      {
+        if (_cellFailures == null)
+          return;
+
+        if (!_cellFailures.TryGetValue(cellDefinition, out var handleableFailures))
+          return;
+
+        foreach (var handleableRowFailure in handleableFailures.Where(e => !e.Handled))
+        {
+          failures.Add(handleableRowFailure.Failure);
+          if (markAsHandled)
+            handleableRowFailure.MarkAsHandled();
+        }
+      }
+
+      private List<HandleableFailure> GetOrCreateListOfCellFailures (IBocColumnDefinitionWithValidationSupport cellDefinition)
+      {
+        _cellFailures ??= new Dictionary<IBocColumnDefinitionWithValidationSupport, List<HandleableFailure>>();
+        if (_cellFailures.TryGetValue(cellDefinition, out var existingFailureList))
+          return existingFailureList;
+
+        var newFailureList = new List<HandleableFailure>();
+        _cellFailures[cellDefinition] = newFailureList;
+        return newFailureList;
+      }
+
+      public bool HasValidationFailures ()
+      {
+        return _rowFailures is { Count: > 0 } || _cellFailures != null && _cellFailures.Values.SelectMany(e => e).Any();
+      }
+    }
+
+    private List<HandleableFailure>? _listValidationFailures;
+    private Dictionary<IBusinessObject, RowFailureContainer>? _rowValidationFailures;
+
+    /// <summary>
+    /// Adds 0..N <see cref="BusinessObjectValidationFailure"/>s that belong to the <see cref="BocList"/> as a whole.
```

Comment:
Descriptions: "Adds something" is enough. mulitplicity doesn't need documetnation unless there's somehting special